drivers/ethernet/eth_stm32_hal_v1.c

```diff
@@ -0,0 +1,711 @@
+	dma_rx_desc = heth->RxFrameInfos.FSRxDesc;
+	/* Set Own bit in Rx descriptors: gives the buffers back to DMA */
+	for (int i = 0; i < heth->RxFrameInfos.SegCount; i++) {
+		dma_rx_desc->Status |= ETH_DMARXDESC_OWN;
+		dma_rx_desc = (ETH_DMADescTypeDef *)
+			(dma_rx_desc->Buffer2NextDescAddr);
+	}
+
+	/* Clear Segment_Count */
+	heth->RxFrameInfos.SegCount = 0;
+
+	/* When Rx Buffer unavailable flag is set: clear it
+	 * and resume reception.
+	 */
+	if ((heth->Instance->DMASR & ETH_DMASR_RBUS) != (uint32_t)RESET) {
+		/* Clear RBUS ETHERNET DMA flag */
+		heth->Instance->DMASR = ETH_DMASR_RBUS;
+		/* Resume DMA reception */
+		heth->Instance->DMARPDR = 0;
+	}
+
+	if (!pkt) {
+		goto out;
+	}
+
+out:
+	if (!pkt) {
+		eth_stats_update_errors_rx(get_iface(dev_data));
+	}
+
+	return pkt;
+}
+
+static void rx_thread(void *arg1, void *unused1, void *unused2)
+{
+	const struct device *dev = (const struct device *)arg1;
+	struct eth_stm32_hal_dev_data *dev_data = dev->data;
+	struct net_if *iface;
+	struct net_pkt *pkt;
+	int res;
+
+	ARG_UNUSED(unused1);
+	ARG_UNUSED(unused2);
+
+	while (1) {
+		res = k_sem_take(&dev_data->rx_int_sem, K_FOREVER);
+		if (res == 0) {
+			/* semaphore taken and receive packets */
+			while ((pkt = eth_rx(dev)) != NULL) {
+				iface = net_pkt_iface(pkt);
+#if defined(CONFIG_NET_DSA_DEPRECATED)
+				iface = dsa_net_recv(iface, &pkt);
+#endif
+				res = net_recv_data(iface, pkt);
+				if (res < 0) {
+					eth_stats_update_errors_rx(
+							net_pkt_iface(pkt));
+					LOG_ERR("Failed to enqueue frame "
+						"into RX queue: %d", res);
+					net_pkt_unref(pkt);
+				}
+			}
+		}
+	}
+}
+
+static void eth_isr(const struct device *dev)
+{
+	struct eth_stm32_hal_dev_data *dev_data = dev->data;
+	ETH_HandleTypeDef *heth = &dev_data->heth;
+
+	HAL_ETH_IRQHandler(heth);
+}
+
+void HAL_ETH_RxCpltCallback(ETH_HandleTypeDef *heth_handle)
+{
+	__ASSERT_NO_MSG(heth_handle != NULL);
+
+	struct eth_stm32_hal_dev_data *dev_data =
+		CONTAINER_OF(heth_handle, struct eth_stm32_hal_dev_data, heth);
+
+	__ASSERT_NO_MSG(dev_data != NULL);
+
+	k_sem_give(&dev_data->rx_int_sem);
+}
+
+static void generate_mac(uint8_t *mac_addr)
+{
+#if defined(ETH_STM32_RANDOM_MAC)
+	/* "zephyr,random-mac-address" is set, generate a random mac address */
+	gen_random_mac(mac_addr, ST_OUI_B0, ST_OUI_B1, ST_OUI_B2);
+#else /* Use user defined mac address */
+	mac_addr[0] = ST_OUI_B0;
+	mac_addr[1] = ST_OUI_B1;
+	mac_addr[2] = ST_OUI_B2;
+#if NODE_HAS_VALID_MAC_ADDR(DT_DRV_INST(0))
+	mac_addr[3] = NODE_MAC_ADDR_OCTET(DT_DRV_INST(0), 3);
+	mac_addr[4] = NODE_MAC_ADDR_OCTET(DT_DRV_INST(0), 4);
+	mac_addr[5] = NODE_MAC_ADDR_OCTET(DT_DRV_INST(0), 5);
+#else
+	uint8_t unique_device_ID_12_bytes[12];
+	uint32_t result_mac_32_bits;
+
+	/* Nothing defined by the user, use device id */
+	hwinfo_get_device_id(unique_device_ID_12_bytes, 12);
+	result_mac_32_bits = crc32_ieee((uint8_t *)unique_device_ID_12_bytes, 12);
+	memcpy(&mac_addr[3], &result_mac_32_bits, 3);
```

Comment:
Needs the same change as in https://github.com/zephyrproject-rtos/zephyr/pull/94638